torch/_library/utils.py

```diff
@@ -96,3 +96,41 @@ def is_functional_schema(schema: Any) -> bool:
     if not schema.returns:
         return False
     return True
+
+
+def is_inplace_aten_op(op: torch._ops.OpOverload):
+    """Check if an op is an inplace aten op, i.e. it mutates and returns the first arg.
+
+    TODO: torchgen/model.py's FunctionSchema.parse is the source of truth for this,
+    but not all PyTorch builds have torchgen (due to the yaml dependency being weird).
+    Figure this out.
+
+    Example: add_(Tensor(a!) x, Tensor y) -> Tensor(a)
+    """
+    if op.namespace != "aten":
+        return False
+    schema = op._schema
+    if not len(schema.returns) == 1:
+        return False
+    if schema.returns[0].alias_info is None:
+        return False
+    alias_set = schema.returns[0].alias_info.after_set
+    if len(alias_set) != 1:
+        return False
+    loc = next(iter(alias_set))
+    if len(schema.arguments) < 1:
+        return False
+    first_arg = schema.arguments[0]
+    if first_arg.alias_info is None:
+        return False
+    if not first_arg.alias_info.is_write:
+        return False
+    alias_set = first_arg.alias_info.after_set
+    if len(alias_set) != 1:
+        return False
+    if loc != next(iter(alias_set)):
+        return False
+    for arg in schema.arguments[1:]:
+        if arg.alias_info is not None:
+            return False
+    return True
```

Comment:
I would prefer to use the existing, well tested path when possible. This can be replaced with

`return schema.is_mutable_input(0) and not any(schema.is_mutable_input(i) for i in range(1, len(arguments)`

and

```
diff --git a/torch/csrc/jit/python/init.cpp b/torch/csrc/jit/python/init.cpp
index 960672e8841..9a11476eb14 100644
--- a/torch/csrc/jit/python/init.cpp
+++ b/torch/csrc/jit/python/init.cpp
@@ -1830,7 +1830,12 @@ void initJITBindings(PyObject* module) {
             return ss.str();
           })
       .def_property_readonly(
-          "is_mutable", [](FunctionSchema& self) { return self.is_mutable(); });
+          "is_mutable", [](FunctionSchema& self) { return self.is_mutable(); })
+      .def(
+          "is_mutable_input", [](FunctionSchema& self, size_t i) { 
+            auto& arg = self.arguments()[i];
+            return self.is_mutable(arg.name());
+      });
```

Reply:
Those aren't good enough: they tell us if the input is being mutated, but they don't tell us if the input is being returned. If we want this to be really robust, then we need to refactor the relevant pieces out of torchgen, which is the source of truth for this.

Reply:
If it's OK with you, could I take this as a follow-up? It's a bit of a pain to refactor torchgen due to internal build reasons.

Reply:
Sure, that's fine with me.